### PR TITLE
CBG-5429: Fix SequenceID String() and Before() for LowSeq/TriggeredBy corner cases 

### PR DESCRIPTION
**Pull request by RIT3shSapata**
CBG-5429

- Fix `SequenceID.String()` and `SequenceID.Before()` for corner cases involving both LowSeq and TriggeredBy, and add a changes feed integration test to verify no regression in backfill behaviour.
- `String()` fix: the old implementation only emitted the LowSeq prefix when LowSeq < Seq. When LowSeq > Seq (which occurs during a backfill-in-progress with active skipped sequences), the LowSeq field was silently dropped, causing the client to receive an incorrect last_seq. On the next request the changes feed would resume from the wrong position, potentially missing backfill documents.
- `Before()` fix — the old implementation routed all comparisons through SafeSequence() (which returns LowSeq when non-zero), ignoring the Seq field entirely when LowSeq is set. This produced wrong ordering for cases such as:                                 
    - 1234:5677:2000.Before(1234::5677) → returned false, expected true                                                              
    - 1234:5677:2000.Before(1300::1100) → returned false, expected true
- Added a test that creates a scenario where skipped sequences (LowSeq mode) and a mid-stream channel access grant (TriggeredBy backfill) are both active simultaneously, verifying that a since value of the form {LowSeq: L, TriggeredBy: T, Seq: S} is correctly serialised, sent back, and processed by the changes feed without missing or duplicating documents.

## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [ ] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## Dependencies (if applicable)
- [ ] Link upstream PRs
- [ ] Update Go module dependencies when merged

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGatewayIntegration/build?delay=0sec)
- [ ]  https://jenkins.sgwdev.com/job/SyncGatewayIntegration/909/
- [x]  https://jenkins.sgwdev.com/job/SyncGatewayIntegration/982/
- [ ]  https://jenkins.sgwdev.com/job/SyncGatewayIntegration/1025/ (failed)
- [x]  https://jenkins.sgwdev.com/job/SyncGatewayIntegration/1028/

### CODE REVIEW

File: rest/changestest/changes_api_test.go [hunk 1]
```diff
@@ -3761,6 +3761,145 @@ func TestOneShotGrantRequestPlusDbConfig(t *testing.T) {
...
+	// already processed earlier (nextSequence=10), this returns immediately.
+	rt.WaitForSequence(7)
+
+	// With seq 7 now in the cache, the changes feed should deliver it via the late-sequence
+	// path. The user still has no DEF access, but the LowSeq advancement itself produces a
+	// result (last_seq moves forward). The last_seq from this response is used as the since
+	// value for the final request below.
+	changes = rt.PostChanges("/{{.keyspace}}/_changes", changesJSON, "sg-user")
+	require.Len(t, changes.Results, 1)
+
```
[the remark this answers is not in the document]
Comment (gregns1) [reply rest/changestest/changes_api_test.go]: false, you will get sequence 9 again here in this scenario when stable sequence moves on

File: rest/changestest/changes_api_test.go [hunk 2]
```diff
@@ -3761,6 +3761,145 @@ func TestOneShotGrantRequestPlusDbConfig(t *testing.T) {
...
+	// Use the last_seq from the post-seq7 response as the since value. This since has LowSeq
+	// set, and after the DEF grant the changes feed will add TriggeredBy, producing a compound
+	// since value {LowSeq: L, TriggeredBy: 10, Seq: S} — the CBG-5429 corner case.
+	changesJSON = fmt.Sprintf(`{"since":"%s"}`, changes.Last_Seq.String())
+
+	// Expect 4 results: the user update (seq 10) plus the 3 DEF backfill docs (seqs 3, 4, 7).
+	// If Before() or String() are broken for the compound since value, backfill docs will be
+	// missed and this assertion will fail.
+	changes = rt.PostChanges("/{{.keyspace}}/_changes", changesJSON, "sg-user")
```
[the remark this answers is not in the document]
Comment (gregns1) [reply rest/changestest/changes_api_test.go]: I think this may be worth doing

**Review by gregns1 — DISMISSED** [review]
I think we have issues with these changes with interrupted/oneshot feeds. I don't think this is effected for continuous feeds but will be effected for continuous feeds on the first iteration of the feed. So lower blast radius. 

## Scenario 1 — continuing backfill skips a doc

| Seq | Channel | Note |
|----|---------|------|
| 1  | (user1) | access: ABC |
| 2,3,4 | DEF | backfill targets |
| 5  | ABC | |
| 6  | — | **gap (skipped)** |
| 7  | ABC | |

1. `GET _changes` as user1 → `last_seq = 5::7`.
2. Grant user1 access to `ABC, DEF` → lands at seq 10.
3. `POST _changes` with `since=5::7&limit=2` (cuts the backfill short) → `last_seq = 5:10:3`.
   *Old code:* `LowSeq(5)` is not `< Seq(3)`, so it emitted `10:3`.
4. Seq 6 arrives; the feed's `lowSequence` moves to 7.
5. `POST _changes` with `since=5:10:3` (no/high limit) → returns only the user doc, `last_seq = 7::10`.

**Result:** Seq 4 is not delivered in the feed. Line 850 in `changes.go` will see that since low seq is != 0 but since low seq != low seq on feed (it’s moved to 7 now) and this does not set since low seq to 0. This stays as 5 here. Then in `changesFeed` `SafeSequence()` calculates safe seq to be low seq, in this case 5 and this leads to sequence 4 above getting missed on the feed. 

## Scenario 2 — grant's fresh backfill is suppressed

| Seq | Channel | Note |
|----|---------|------|
| 1  | (user1) | access: ABC |
| 2  | DEF | backfill target |
| 3  | GHI | backfill target |
| 4  | (grant) | user1 → ABC, DEF |
| 5  | (grant) | user1 → ABC, DEF, GHI |
| 6,7 | ABC | |
| 8  | — | **gap (skipped)** |
| 9  | ABC | |

1. `POST _changes` with `limit=1` (cuts the backfill short) → `last_seq = 7:4:2`.
2. Seq 8 arrives; the feed's `lowSequence` moves off 7.
3. `POST _changes` with `since=7:4:2`.

**Result:** GHI's backfill doc (seq 3) is not delivered in feed. On this first iteration `isNewChannel` is false (`changedChannels` is only populated on later iterations of a waiting feed), so the decision to backfill falls to `backfillRequired` at changes.go:929
So `seqAdded` at for GHI will be seq 5:
So `seqAddedAt` > 1 -> true 
But `7:4:2.Before(5)` -> false 
`backfillRequired` is false

I think overall it may be worth discussing as a team this PR in sync up.

File: db/sequence_id.go [hunk 3]
```diff
@@ -52,22 +52,34 @@ func (s SequenceID) String() string {
...
 func (s SequenceID) intSeqToString() string {
-
-	if s.LowSeq > 0 && s.LowSeq < s.Seq {
+	// LowSeq is omitted from the output here for two independent reasons: it's zero (there's
+	// nothing to report), or it's stale - greater than Seq. LowSeq is stale when this entry is a
+	// previously-skipped sequence being delivered after the feed's lowest contiguous sequence has
+	// already moved past it, so including it here would misrepresent this entry's position.
+	// TriggeredBy, if set, reflects an in-progress channel backfill and is unrelated to which of
+	// those two reasons applies.
+	if s.LowSeq == 0 || s.Seq < s.LowSeq {
```
Comment (adamcfraser) [inline db/sequence_id.go]: "stale" is the wrong way to describe this scenario.  If LowSeq is greater than s.Seq, it represents a case where the sequence of the revision being sent (s.Seq) is older than the cache's stable sequence (LowSeq), and so it's not necessary to include the stable sequence information.  
The low (stable) sequence is only required for revisions with sequences higher than the stable sequence, to keep track of the fact that we're sending a sequence later than the stable sequence, and so there are potentially sequences n in the range s.LowSeq < n < s.Seq that the client hasn't yet received.

File: db/sequence_id_test.go [hunk 4]
```diff
@@ -185,3 +210,46 @@ func TestCompareSequenceIDsLowSeq(t *testing.T) {
...
+		},
+		{
+			name:   "compound sequence with lowSeq, triggeredBy seq and seq",
+			seq:    SequenceID{LowSeq: 100, TriggeredBy: 105, Seq: 103},
+			seqStr: "100:105:103",
+		},
+		{
+			name:   "compound sequence with lowSeq, triggeredBy seq and seq and seq < lowSeq",
+			seq:    SequenceID{LowSeq: 100, TriggeredBy: 105, Seq: 90},
+			seqStr: "105:90",
```
Comment (adamcfraser) [inline db/sequence_id_test.go]: This is incorrect - this should be 100:105:90.

105:90 indicates that the changes feed is currently at sequence 105, and there's a channel grant at sequence 105 that has triggered a backfill.  Sequence 90 is being sent as part of that backfill, but the overall feed is still at sequence 105.  Omitting lowSeq here loses track of the fact that the client hasn't been sent all sequences between 100 and 105.

LowSeq should be included whenever lowSeq is lower than the overall position of the feed.  For simple sequences (no triggered by), that's just when lowSeq < Seq.  But if there's a backfill, the position of the feed is TriggeredBy, so you need to compare lowSeq < TriggeredBy.

File: db/sequence_id.go (hunk 3 above)
Comment (adamcfraser) [inline db/sequence_id.go]: The evaluation to determine whether lowSeq should be sent (s.Seq < s.LowSeq) is only valid when triggeredBy=0.  If triggeredBy is non-zero the evaluation should be (s.Seq < s.TriggeredBy).  See the notes in the unit test below.

File: db/sequence_id_test.go (hunk 4 above)
Comment (adamcfraser) [reply db/sequence_id_test.go]: I added another table to the spreadsheet for all the String() permutations, have a look and add them as use cases for this test.

**Review by adamcfraser — CHANGES_REQUESTED** [review]
A few changes needed.

File: db/sequence_id.go [hunk 5]
```diff
@@ -53,21 +53,57 @@ func (s SequenceID) String() string {
 }
 
 func (s SequenceID) intSeqToString() string {
+	seqStrHasTriggeredBy := false
+	seqStr := fmt.Sprintf("%d", s.Seq)
```
Comment (adamcfraser) [inline db/sequence_id.go]: Sprintf is the (relatively) expensive operation in this function, so we should avoid doing it multiple times.  Right now if triggered by and low seq are both defined this function is doing three separate Sprintf (discarding the first two) - we should refactor to avoid that.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: db/sequence_id.go (hunk 5 above)
Comment (RIT3shSapata) [reply db/sequence_id.go]: refactored the function

File: db/sequence_id.go [hunk 6]
```diff
@@ -53,21 +53,21 @@ func (s SequenceID) String() string {
 }
 
 func (s SequenceID) intSeqToString() string {
-
-	if s.LowSeq > 0 && s.LowSeq < s.Seq {
-		if s.TriggeredBy > 0 {
+	backfillActive := s.TriggeredBy > 0 && s.Seq < s.TriggeredBy
```
Comment (adamcfraser) [inline db/sequence_id.go:61]: intSeqToString() now does the work to drop TriggeredBy in the String representation when Seq >= TriggeredBy.  Previously that was the responsibility of the caller.  This might be reasonable, but I think a comment describing the cases where intSeqToString() is going to omit a non-zero value would be useful.  I think those are:
- omit LowSeq when it's higher than the current sequence (non-backfill)
- omit LowSeq when it's higher than the triggeredBy sequence (backfill)
- omit TriggeredBy when it's higher than the current sequence

File: rest/audit_test.go [hunk 7]
```diff
@@ -1355,7 +1355,7 @@ func TestAuditChangesFeedStart(t *testing.T) {
 				expectedFields: map[string]any{
 					base.AuditFieldFeedType:    "normal",
 					base.AuditFieldIncludeDocs: true,
-					base.AuditFieldSince:       "5:10",
+					base.AuditFieldSince:       "10",
```
Comment (adamcfraser) [inline rest/audit_test.go]: I think the intention of this test was to validate non-integer since values were properly emitted - it may be better to change the test to `since=10:5` to preserve this.

Comment (adamcfraser) [reply rest/audit_test.go]: More generally, though, this change in behaviour feels like it may be a concern.  For the purposes of auditing (and logging in general), we'd expect to emit exactly what was provided by the client, and not apply any of the logic in intSeqToString.
It's true that if SGW isn't emitting sequences in invalid formats like `5:10` then we shouldn't expect to be receiving those either, but this type of information might be useful (e.g. to triage a client-side bug).  
I think this could be handled in a followup ticket - can you file one to ensure that in places we're logging a since value received from a client (whether blip or REST), we're logging that exactly as received?

File: rest/changestest/changes_backfill_lowseq_test.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+	// Use the last_seq from the post-seq7 response as the since value. This since has LowSeq
+	// set, and after the DEF grant the changes feed will add TriggeredBy, producing a compound
+	// since value {LowSeq: L, TriggeredBy: 10, Seq: S} — the CBG-5429 corner case.
+	changesJSON = fmt.Sprintf(`{"since":"%s"}`, changes.Last_Seq.String())
+
+	// Expect 4 results: the user update (seq 10) plus the 3 DEF backfill docs (seqs 3, 4, 7).
+	// If Before() or String() are broken for the compound since value, backfill docs will be
+	// missed and this assertion will fail.
+	changes = rt.PostChanges("/{{.keyspace}}/_changes", changesJSON, "sg-user")
+	require.Len(t, changes.Results, 4)
```
Comment (adamcfraser) [inline rest/changestest/changes_backfill_lowseq_test.go:343]: It would be useful to perform and additional changes request with the lastSeq from this call (7:10) to ensure doesn't retrigger backfill (should return zero results if the lowSeq hasn't changed.).

After that, you should write sequence 8 in channel DEF, and make sure it is sent to the client even though the backfill had completed.  (I think we expect the entire backfill for DEF to be sent again in this case).

If you've got this test elsewhere that's fine, but otherwise can you add it here?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
Functional changes look good, but a few additional test cases would be useful to ensure everything's working as expected, and to avoid future regressions.

File: rest/audit_test.go (hunk 7 above)
Comment (RIT3shSapata) [reply rest/audit_test.go]: changed it, filed:
https://jira.issues.couchbase.com/browse/CBG-5670

**Review by adamcfraser — APPROVED** [review]
LGTM!
